builder/api/src/main/java/io/helidon/builder/api/ProvidedUtil.java

```diff
@@ -0,0 +1,192 @@
+        for (ConfiguredService service : configuredServices) {
+            ConfiguredProvider<T> provider = allProvidersByType.get(service.type());
+            if (provider == null) {
+                throw new ConfigException("Unknown provider configured. Expecting a provider with type \"" + service.type()
+                                                  + "\", but only the following providers are supported: "
+                                                  + allProvidersByType.keySet() + ", "
+                                                  + "provider interface: " + providerType.getName()
+                                                  + ", configured service: " + configType.getName());
+            }
+            unusedProvidersByType.remove(service.type());
+            if (service.enabled()) {
+                result.add(provider.create(service.serviceConfig(), service.name()));
+            }
+        }
+
+        // then (if desired) add the rest
+        if (allFromServiceLoader) {
+            unusedProvidersByType.values()
+                    .forEach(it -> result.add(it.create(Config.empty(), it.configKey())));
+        }
+
+        return result;
+    }
+
+    private static ConfiguredService configuredService(Config serviceConfig, boolean isList) {
+        if (isList) {
+            // order is significant
+            String type = serviceConfig.get("type").asString().orElse(null);
+            String name = type;
+            boolean enabled = serviceConfig.get("enabled").asBoolean().orElse(true);
+
+            Config usedConfig = serviceConfig;
+            if (type == null) {
+                // nested approach (we are on the first node of the list, we need to go deeper)
+                List<Config> configs = serviceConfig.asNodeList().orElseGet(List::of);
+                if (configs.size() != 1) {
+                    throw new ConfigException(
+                            "Service provider configuration defined as a list must have a single node that is the type, "
+                                    + "with children containing the provider configuration. Failed on: " + serviceConfig.key());
+                }
+                usedConfig = configs.get(0);
+                name = usedConfig.name();
+                type = usedConfig.get("type").asString().orElse(name);
+                enabled = usedConfig.get("enabled").asBoolean().orElse(enabled);
+            }
+            return new ConfiguredService(usedConfig, type, name, enabled);
+        }
+        // just collect each node, order will be determined by weight
+
+        String name = serviceConfig.name();
+        String type = serviceConfig.get("type").asString().orElse(name);
+        boolean enabled = serviceConfig.get("enabled").asBoolean().orElse(true);
+
+        return new ConfiguredService(serviceConfig, type, name, enabled);
+    }
+
+    private record ConfiguredService(Config serviceConfig, String type, String name, boolean enabled) {
```

Comment:
??

Reply:
Not sure what is your problem here, please explain. You have marked a record used internally to have multiple return values from a method. What is the issue?